src/Npgsql/Internal/NpgsqlConnector.cs

```diff
@@ -2624,10 +2624,11 @@ internal async Task<bool> Wait(bool async, int timeout, CancellationToken cancel
         var keepaliveMs = Settings.KeepAlive * 1000;
         while (true)
         {
-            cancellationToken.ThrowIfCancellationRequested();
-
             var timeoutForKeepalive = _isKeepAliveEnabled && (timeout <= 0 || keepaliveMs < timeout);
             ReadBuffer.Timeout = TimeSpan.FromMilliseconds(timeoutForKeepalive ? keepaliveMs : timeout);
+            // We should set the timeout before checking the token as otherwise we're racing with the callback registered on cancellationToken
+            // Which can also change the timeout
+            cancellationToken.ThrowIfCancellationRequested();
```

Comment:
I don't think this is guaranteed to help. `cancellationToken.ThrowIfCancellationRequested()` only provides a read acquire barrier, and I don't see any other barriers as a part of setting `ReadBuffer.Timeout`, so the runtime is free to move the call to `ThrowIfCancellationRequested()` before the write to the timeout.

Reply:
Isn't according to [memory model](https://github.com/dotnet/runtime/blob/main/docs/design/specs/Memory-model.md#order-of-memory-operations) acquire barrier guarantees that any reads and writes can only happen after that read?

> Volatile reads have "acquire semantics" - no read or write that is later in the program order may be speculatively executed ahead of a volatile read.

And from my understanding that's relevant if and only if `ThrowIfCancellationRequested` is inlined, as otherwise the runtime can't guarantee that it doesn't have side effects regarding `ReadBuffer`, so it can't move it after `ReadBuffer.Timeout =`.

Reply:
You won't see an x64 cpu do so but on arm I suppose it could happen... 
 
The JIT likely won't move the code as the write to the timeout and the read of the cts reference (until the `if canceled` branch) should be part of the same basic block, but that's mostly an implementation detail.

When I set these multi threaded timeout values in Slon I have an `Interlocked.MemoryBarrier()` after it, mostly to make sure it propagates to timer threads in a timely fashion. It would be enough here too (assuming this solves the deadlock, I haven't looked too closely).

I don't see the volatile read @WizardBrony referred to in ThrowIfCancellationRequested. Writing an object reference forces a barrier and is guaranteed to be atomic but reading an object reference has no barriers in .NET (unlike many low latency GCs that require them).

Reply:
I don’t think I follow. My point was that this change depends on the write to ‘ReadBuffer.Timeout’ happening before the token is checked, and according to the C# memory model, it is valid for the runtime to reorder the write after the read. Please let me know if I am missing something.

Reply:
> I don't see the volatile read @WizardBrony referred to in ThrowIfCancellationRequested.

https://github.com/dotnet/runtime/blob/040cbe276907174316e2cc07b35814b3069874a6/src/libraries/System.Private.CoreLib/src/System/Threading/CancellationTokenSource.cs#L73

And `_state_` is marked as `volatile`.

Reply:
@vonzshik Correct. And that volatile read does not prevent `ReadBuffer.Timeout` from floating below `ThrowIfCancellationRequested`.

Reply:
As @NinoFloris mentioned though, `Interlocked.MemoryBarrier()` would prevent that.

Reply:
@WizardBrony I'm mostly agreeing with you, I just don't see that acquire barrier that you mentioned (EDIT: ah ok it's in the CTS itself).

Either way to expand a bit on what I said:

Starting from the top, roslyn won't reorder these operations as calls are potentially side effecting, and roslyn does no inlining.

Onto the JIT:

> The JIT likely won't move the code as the write to the timeout and the read of the cts reference (until the if canceled branch) should be part of the same basic block, but that's mostly an implementation detail.

It's an implementation detail of RyuJIT, and as you mentioned the memory model has no such restriction.

If we do take that implementation detail as our model then there is only one place left that could reorder things, the cpu. 

For x86 and x64 each store has release semantics by default, so if the JIT didn't move the read and the cpu won't either it would all work correctly without adding explicit barriers *on x64*. On arm however there is again no such guarantee. 

So, to be correct here you'd want at least a Volatile.Write of the timeout, but ideally an Interlocked.MemoryBarrier. The reason for wanting the barrier is that you reduce the chances the race with the timer thread happens, as the barrier ensures the write is globally visible before continuing (i.e. store buffers were flushed). Volatile.Writes only have release semantics, which only prevent later operations from being observable before the write can be observed.

Reply:
OK, it seems like my brain was too tired today and looked at the previous implementation (before this pr), where there is no reordering.
But, looking at the code again, there is an issue where `NpgsqlReadBuffer.Timeout` isn't thread safe, so it's possible that we can get concurrent writes from a callback registered on `CancellationToken` and the method itself. This wouldn't be solved by `Interlocked.MemoryBarrier`, so I added a lock there. This ensures that at the moment we have a lock either cancellation went through completely (at which point we get OCE), or it didn't, and then the callback will change timeout to -1 and we'll get an error immediately on read.

Reply:
Reading is fine, it just a single read from a `TimeSpan` field, which is atomic on x64. And on x86 it's technically possible to get torn read, but to even accomplish that you have to have concurrent query cancellation on the moment this check happens, and even then it's probably doesn't matter that much since we do not care about particular value, just whether it's 0 or not.
The main problem is writing as it has a few checks and writes multiple other fields.

Reply:
@vonzshik What about these writes?
https://github.com/npgsql/npgsql/blob/5975-wait-cancellation-token-cancel-deadlock-fix/src/Npgsql/Internal/NpgsqlConnector.cs#L2589
https://github.com/npgsql/npgsql/blob/5975-wait-cancellation-token-cancel-deadlock-fix/src/Npgsql/Internal/NpgsqlConnector.cs#L2490

Reply:
The first one (keepalive) is performed whenever we do not have an active query, so we're not racing there (lock over `SyncObj` should guarantee that).
The second one is a bit more tricky. We call that method just before we start executing a query/start an action, so it's supposed to be the only one being able to change things around. But there is still a miniscule possibility of concurrent write due to `StartCancellableOperation`, which registers cancellation token callback. Overall a chance of something like that ever happening is extremely miniscule since cancellation callback has to do a lot of work before even being able to try and cancel a query, but just in case I did move it after changing the timeout.

Reply:
Hey @vonzshik, just a side-note. Right before you replied to me I had deleted my comment asking about the writes to `ReadBuffer.Timeout` because I had come to the same conclusions you did, but also for the write in `DoStartUserAction` I realized that even if there was a race with the cancellation callback it wouldn’t really cause an issue for `Wait`, so I figured it was off-topic for this PR. But I appreciate you being thorough and fixing it for the non-`Wait` code paths anyway. Thanks for all your work on this.